spec.bs

```diff
@@ -5863,6 +5832,204 @@ an [=origin=] |topLevelOrigin|, and a [=string=] |slotSizeQueryParam|:
+    |adComponentRenderURLs|.
+1. If |experimentGroupId| is not null:
+  1. [=list/Append=] "`&experimentGroupId=`" to |queryParamsList|.
+  1. [=list/Append=] [=serialize an integer|serialized=] |experimentGroupId| to |queryParamsList|.
+1. Set |signalsUrl|'s [=url/query=] to the result of [=string/concatenating=] |queryParamsList|.
+1. return |signalsUrl|.
+
+</div>
+
+<div algorithm>
+To <dfn>build batched trusted scoring signals url</dfn> given a non-empty [=list=] of [=trusted
+scoring signals requests=] |entriesToBatch|:
+
+1. Let |firstEntry| be |entriesToBatch|[0].
+1. Let |renderURLs| be an empty [=ordered set=] of [=strings=].
+1. Let |adComponentRenderURLs| be an empty [=ordered set=] of [=strings=]
+  |adComponentRenderURLs|
+1. [=list/For each=] |entry| of |entriesToBatch|:
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/base URL=] is equal to
+    |firstEntry|'s [=trusted scoring signals request/base URL=].
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/seller experiment group id=] is
+    equal to |firstEntry|'s [=trusted scoring signals request/seller experiment group id=].
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/top level origin=] is equal to
+    |firstEntry|'s [=trusted scoring signals request/top level origin=].
+  1. [=Assert=] that |entry|'s [=trusted scoring signals request/seller script fetcher=] is equal to
+    |firstEntry|'s [=trusted scoring signals request/seller script fetcher=].
+  1. [=set/Append=] |entry|'s [=trusted scoring signals request/render URL=] to |renderURLs|.
+  1. [=list/Extend=] |adComponentRenderURLs| with |entry|'s [=trusted scoring signals request/
+    ad component URLs=].
+1. Return the result of [=building trusted scoring signals url=] with |firstEntry|'s [=trusted
+  scoring signals request/base URL=], |renderURLs|, |adComponentRenderURLs|, |firstEntry|'s
+  [=trusted scoring signals request/seller experiment group id=],
+  |firstEntry|'s [=trusted scoring signals request/top level origin=].
+
+</div>
+
+<div algorithm>
+To <dfn>check if trusted scoring signals batch honors URL length limit</dfn> given a
+[=trusted scoring signals batcher=] |batcher| and a non-empty [=list=] of [=trusted
+scoring signals requests=] |entriesToBatch|:
+
+1. If the [=list/size=] of |entriesToBatch| is 1, return true.
+1. If |batcher|'s [=trusted scoring signals batcher/url length limit=] is 0, return true.
+1. Return whether the [=string/length=] of the result of [=building batched trusted scoring signals
+  url=] for |entriesToBatch| is &le; |batcher|'s [=trusted scoring signals batcher/url length
+  limit=].
+
+</div>
+
+<div algorithm>
+
+To <dfn>batch and fetch trusted scoring signals</dfn> given a [=trusted scoring signals batcher=]
+|batcher|:
+
+1. Until this object is no longer needed:
+  1. Wait until |batcher|'s [=trusted scoring signals batcher/request queue=] is no longer empty or
+    some heuristically chosen amount of time has passed.

```

Comment:
No, because this may be empty by the "request map" field might not be.

... Though the distinction between request queue and request map only exists because I have no idea on how to express an atomic update to "request map", so the ideal solution would be to only have request map, have things atomically inserted into it, and then the condition can be as you said.

Reply:
If this is intended to trigger when the queue is empty, then "Wait until |batcher|'s [=trusted scoring signals batcher/request queue=] is no longer empty" doesn't work alone.  I had assumed this was an exclusive or.  I still think we want to rephrase this.

Reply:
So I read this as do one or the other, not having two different triggers in the same implementation.  i.e.:

option1:

void ThreadFunc() {
  while (running()) {
    WaitForNonEmpty();
    DoStuff();
  }
}

or

void ThreadFunc() {
  while (running()) {
    WaitForHeuristic();
    DoStuff();
  }
}

As opposed to:

void ThreadFunc() {
  while (running()) {
    MagicallyCombineWaits(&WaitForNonEmpty, &WaitForHeuristic);
    DoStuff();
  }
}

I think we should make clear it's the latter.  The heuristic wait is a mandatory part of the algorithm (at least if the other heuristic can leave map non-empty)

Reply:
Thanks, yeah, that would be an issue. Do you perhaps have suggestions on clearer phrasing?

Reply:
Maybe:

Wait until one of the following is true:
  1. |batcher|'s [=trusted scoring signals batcher/request queue=] is no longer empty.
  1. |batcher|'s [=trusted scoring signals batcher/request map=] is non-empty and some heuristically chosen amount of time has passed.

Reply:
Thanks, this is indeed much better; done, though I used "at least one"